<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  
  <link rel="shortcut icon" href="../img/favicon.ico">
  <title>Tutorial - GeneSeekr Documentation</title>
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  <link rel="stylesheet" href="../css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../css/theme_extra.css" type="text/css" />
  <link rel="stylesheet" href="../css/highlight.css">
  
  <script>
    // Current page data
    var mkdocs_page_name = "Tutorial";
    var mkdocs_page_input_path = "tutorial.md";
    var mkdocs_page_url = "/tutorial/";
  </script>
  
  <script src="../js/jquery-2.1.1.min.js"></script>
  <script src="../js/modernizr-2.8.3.min.js"></script>
  <script type="text/javascript" src="../js/highlight.pack.js"></script> 
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
      <div class="wy-side-nav-search">
        <a href=".." class="icon icon-home"> GeneSeekr Documentation</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
	<ul class="current">
	  
          
            <li class="toctree-l1">
		
    <a class="" href="..">Home</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../installation/">Installation</a>
	    </li>
          
            <li class="toctree-l1 current">
		
    <a class="current" href="./">Tutorial</a>
    <ul class="subnav">
            
    <li class="toctree-l2"><a href="#tutorial">Tutorial</a></li>
    
        <ul>
        
            <li><a class="toctree-l3" href="#basic-settings">Basic settings</a></li>
        
            <li><a class="toctree-l3" href="#commands">Commands</a></li>
        
            <li><a class="toctree-l3" href="#arguments">Arguments</a></li>
        
            <li><a class="toctree-l3" href="#optional-arguments">Optional Arguments</a></li>
        
            <li><a class="toctree-l3" href="#additional-optional-arguments">Additional Optional Arguments</a></li>
        
        </ul>
    

    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../tests/">Tests</a>
	    </li>
          
        </ul>
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="..">GeneSeekr Documentation</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="..">Docs</a> &raquo;</li>
    
      
    
    <li>Tutorial</li>
    <li class="wy-breadcrumbs-aside">
      
        <a href="https://github.com/OLC-Bioinformatics/GeneSeekr/edit/master/docs/tutorial.md"
          class="icon icon-github"> Edit on GitHub</a>
      
    </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <h2 id="tutorial">Tutorial</h2>
<h3 id="basic-settings">Basic settings</h3>
<h3 id="commands">Commands</h3>
<p>GeneSeekr requires the desired BLAST program to be specified. The
acceptable options are:</p>
<pre><code>* blastn (nucleotide:nucleotide)
* blastp (protein:protein)
* blastx (translated nucleotide:protein)
* tblastn (protein:translated nucleotide)
* tblastx (translated nucleotide: translated nucleotide)
</code></pre>
<h3 id="arguments">Arguments</h3>
<p>The pipeline is designed to run with a minimum of three arguments:</p>
<pre><code>* path to sequence data (-s)
* path to target database (-t)
* path in which report folder is to be created (-r)
</code></pre>
<p>The following command will run geneseekr using BLASTn on the supplied sequences and database</p>
<pre><code>geneseekr blastn -s /path/to/sequences -t /path/to/targets -r /path/to/reports

</code></pre>

<h3 id="optional-arguments">Optional Arguments</h3>
<p>There are a number of optional arguments than can be supplied to the geneseekr script</p>
<pre><code>  Options:
  -s, --sequencepath TEXT   Specify input fasta folder  [required]
  -t, --targetpath TEXT     Specify folder of targets  [required]
  -r, --reportpath TEXT     Specify output folder for csv  [required]
  -c, --cutoff INTEGER      Minumum sequence identity threshold for a match to
                            be reported
  -e, --evalue TEXT         Minimum evalue to use for BLAST analyses
  -n, --numthreads INTEGER  Specify number of threads
  -a, --align               Optionally output alignments of genes with less
                            than 100% identity to reference genes. This
                            alignment will use amino acid sequences for both
                            query and reference
  -u, --unique              Do not report multiple hits at the same location
                            in a contig. Instead, store the best hit, and
                            ignore the rest
  --version                 Show the version and exit.
  -h, --help                Show this message and exit.
</code></pre>

<h3 id="additional-optional-arguments">Additional Optional Arguments</h3>
<p>If you use a BLAST program that employs a nucleotide database(BLASTn, tBLASTn, or tBLASTx) the following arguments
are available.</p>
<pre><code>  -R, --resfinder           Perform ResFinder-like analyses
  -V, --virulencefinder     Perform VirulenceFinder-like analyses
</code></pre>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../tests/" class="btn btn-neutral float-right" title="Tests">Next <span class="icon icon-circle-arrow-right"></span></a>
      
      
        <a href="../installation/" class="btn btn-neutral" title="Installation"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="http://www.mkdocs.org">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" style="cursor: pointer">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
          <a href="https://github.com/OLC-Bioinformatics/GeneSeekr/" class="fa fa-github" style="float: left; color: #fcfcfc"> GitHub</a>
      
      
        <span><a href="../installation/" style="color: #fcfcfc;">&laquo; Previous</a></span>
      
      
        <span style="margin-left: 15px"><a href="../tests/" style="color: #fcfcfc">Next &raquo;</a></span>
      
    </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme.js"></script>
      <script src="../search/require.js"></script>
      <script src="../search/search.js"></script>

</body>
</html>
